bootstrap/src/sun/nio/ch/lincheck/Injections.java

```diff
@@ -45,6 +50,101 @@ public class Injections {
+     * Thread local variable storing the thread descriptor for each thread.
+     *
+     * NOTE: as an optimization, for the `TestThread` instances,
+     *   we avoid lookup via `ThreadLocal` by instead storing
+     *   the thread descriptor in a field of the thread object itself.
+     */
+    private static final ThreadLocal<ThreadDescriptor> threadDescriptor =
+        ThreadLocal.withInitial(() -> null);
+
+    /*
+     * A global map storing a thread descriptor for each thread object.
+     *
+     * The map should be concurrent, use identity hash code,
+     * and do not prevent garbage collection of thread objects (i.e., use weak keys).
+     * However, because there is no `ConcurrentWeakIdentityHashMap` out of the box in
+     * the java standard library, we use a few tricks to achieve a similar effect.
+     *
+     * - We use `ConcurrentHashMap` to guarantee thread safety.
+     * - We use identity hash codes of the thread objects as keys.
+     * - To accommodate for potential hash code collisions, we store lists of thread descriptors as values.
+     * - These lists use copy-on-write strategy when a new descriptor is added to avoid race conditions.
+     * - Thread descriptors store weak references to thread objects, and thus do not prevent their garbage collection.
+     *
+     * TODO: although the garbage collection of thread objects is not prevented thanks to weak references,
+     *   the hash map entries themself (key - identity hash code, and value - `ThreadDescriptor` object)
+     *   would not be garbage collected, potentially creating a memory leak.
+     *   However, we expect that typical programs would create only a bounded number of threads during
+     *   their whole lifetime, and thus this map also would only occupy a bounded amount of memory.
+     *   Still, it would be good to eventually implement proper periodic clean-up of the map
+     *   to remove obsolete entries.
+     */
+    private static final ConcurrentHashMap<Integer, ArrayList<ThreadDescriptor>> threadDescriptorsMap =
+        new ConcurrentHashMap<Integer, ArrayList<ThreadDescriptor>>();
+
+    public static ThreadDescriptor getCurrentThreadDescriptor() {
+        Thread thread = Thread.currentThread();
+        if (thread instanceof TestThread) {
+            return ((TestThread) thread).descriptor;
+        }
+        return threadDescriptor.get();
+    }
+
+    public static void setCurrentThreadDescriptor(ThreadDescriptor descriptor) {
+        Thread thread = Thread.currentThread();
+        if (thread instanceof TestThread) {
+            return;
+        }
+        threadDescriptor.set(descriptor);
+    }
+
+    public static ThreadDescriptor getThreadDescriptor(Thread thread) {
+        if (thread instanceof TestThread) {
+            return ((TestThread) thread).descriptor;
+        }
+        int hashCode = System.identityHashCode(thread);
+        ArrayList<ThreadDescriptor> threadDescriptors = threadDescriptorsMap.get(hashCode);
```

Comment:
Can we use a weak hash map instead?

Reply:
I am afraid no, because we need concurrent identity hash map with weak keys, and there is no such data structure out-of-the-box in the Java standard library. 

Please also have a look at the comment before `threadDescriptorsMap` field declaration --- it explains the reason and the proposed solution.

Reply:
Would using `Collections.synchronizedMap(...)` solve the issue?

Reply:
`Collections.synchronizedMap(...)` would not fully solve the problem. It would make the collection thread-safe, but `WeakIdentityHashMap` is still not a thing in Java stdlib (remember we need thread-safe weak identity hash-map).

Also, there is a method to lookup for a descriptor of the given thread:
```
public static ThreadDescriptor getThreadDescriptor(Thread thread)
``` 

which uses `threadDescriptorsMap.get`. 

Forcing the descriptor lookup function to acquire a lock on each lookup (instead of using concurrent hash map), may in theory have a performance penalty.

Reply:
Fixed, changed to `WeakIdentityHashMap` + `Collections.synchronizedMap(...)`